<a href="https://colab.research.google.com/github/Mathildeholst/Case/blob/main/OTARI_Case.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Load og hent data til Colab

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd

folder = "/content/drive/MyDrive/OTARI_case"

finance_path = f"{folder}/Finance.csv"
companies_path = f"{folder}/Companies.csv"

In [ ]:
finance = pd.read_csv(finance_path)
companies = pd.read_csv(companies_path)

companies.columns = companies.columns.str.strip()
finance.columns = finance.columns.str.strip()

#Step 1: Data understanding & quality

Se kolonnenavne, struktur og hvordan værdierne ser ud

In [4]:
import pandas as pd

finance_path = "/content/drive/MyDrive/OTARI_case/Finance.csv"
companies_path = "/content/drive/MyDrive/OTARI_case/Companies.csv"

finance_sample = pd.read_csv(finance_path, nrows=10)
companies_sample = pd.read_csv(companies_path, nrows=10)

print("FINANCE")
display(finance_sample)

print("\nCOMPANIES")
display(companies_sample)

FINANCE


,CVR,LocalName,Value,Context,SlutDato,InstantDate,DocType,DKLabel
0,32945856,AverageNumberOfEmployees,1,D0,2025-06-30,NaN,AARSRAPPORT,Antal ansatte
1,32945856,GrossProfitLoss,-17595,D0,2025-06-30,NaN,AARSRAPPORT,Brutto fortjeneste
2,32945856,OtherFinanceIncome,36365,D0,2025-06-30,NaN,AARSRAPPORT,Andre finansielle indtægter
3,32945856,OtherFinanceExpenses,0,D0,2025-06-30,NaN,AARSRAPPORT,Andre finansielle udgifter
4,32945856,ProfitLossFromOrdinaryActivitiesBeforeTax,18770,D0,2025-06-30,NaN,AARSRAPPORT,Resultat før skat
5,32945856,ProfitLoss,14986,D0,2025-06-30,NaN,AARSRAPPORT,Resultat
6,32945856,Assets,4355073,I0,2025-06-30,2025-06-30,AARSRAPPORT,Balancesum
7,32945856,Equity,4109281,I0,2025-06-30,2025-06-30,AARSRAPPORT,Egenkapital
8,32945856,CashAndCashEquivalents,4155443,I0,2025-06-30,2025-06-30,AARSRAPPORT,Likvide beholdninger
9,32945856,LongtermLiabilitiesOtherThanProvisions,3784,I0,2025-06-30,2025-06-30,AARSRAPPORT,Langsigtede forpligtelser



COMPANIES


,UnitNumber,kommuneNavn,Postnummer,Navn,Branche,Cvr,Status
0,4004306483,VEJEN,6600,PHØNIX BELÆGNINGSTEKNIK A/S,NaN,20877014,OPLØSTEFTERFUSION
1,4004306484,NaN,3730,BALTIC BORN TECH A/S,NaN,76164010,OPLØSTEFTERFRIVILLIGLIKVIDATION
2,4004306485,BRØNDBY,2600,COMPUTERVISION DANMARK A/S,NaN,66616215,OPLØSTEFTERFUSION
3,4004306486,ODENSE,5250,TØMRERMESTER CHR. ELLEGAARD ANDERSEN A/S,NaN,76999910,OPLØSTEFTERFRIVILLIGLIKVIDATION
4,4004306487,HILLERØD,3400,"DATA-TIME, HILLERØD A/S",NaN,73333016,OPLØSTEFTERKONKURS
5,4004306488,HØRSHOLM,2960,SEAWEAR A/S,NaN,70646013,OPLØSTEFTERFRIVILLIGLIKVIDATION
6,4004306489,NaN,8900,"BENT JENSEN TAGDÆKNING, RANDERS, A/S",NaN,17670107,OPLØSTEFTERKONKURS
7,4004306490,KØBENHAVN,1019,"AKTIESELSKABET ""DET KØBENHAVNSKE EJENDOMS-SOCI...",NaN,14707700,OPLØSTEFTERKONKURS
8,4004306491,ODENSE,5100,A/S 21/2 1988 (BOOKER),NaN,39728818,OPLØSTEFTERFUSION
9,4004306492,FAKSE,4640,A/S BRYGGERIET ODIN,NaN,45726916,OPLØSTEFTERFUSION


###Companies

In [5]:
companies = pd.read_csv(companies_path)

print("Shape:", companies.shape)

print("\nDatatypes:")
display(companies.dtypes)

print("\nMissing values:")
display(
    companies.isna()
    .sum()
    .to_frame("Missing")
    .assign(Percent=lambda x: (x["Missing"] / len(companies) * 100).round(2))
)

print("\nUnique values:")
display(companies.nunique().to_frame("Unique"))

Shape: (2250983, 7)

Datatypes:


,0
UnitNumber,int64
kommuneNavn,object
Postnummer,float64
Navn,object
Branche,object
Cvr,int64
Status,object



Missing values:


,Missing,Percent
UnitNumber,0,0.00
kommuneNavn,58501,2.60
Postnummer,473,0.02
Navn,501,0.02
Branche,67894,3.02
Cvr,0,0.00
Status,1,0.00



Unique values:


,Unique
UnitNumber,2250983
kommuneNavn,330
Postnummer,1196
Navn,2179890
Branche,1754
Cvr,2250983
Status,22


In [8]:
companies.columns = companies.columns.str.strip()

print(companies.columns.tolist())

['UnitNumber', 'kommuneNavn', 'Postnummer', 'Navn', 'Branche', 'Cvr', 'Status']


In [10]:
print("COMPANY STATUS")
display(
    companies["Status"]
    .value_counts(dropna=False)
    .to_frame("Count")
)

print("\nMOST COMMON INDUSTRIES")
display(
    companies["Branche"]
    .value_counts(dropna=False)
    .head(20)
    .to_frame("Count")
)

print("\nMOST COMMON MUNICIPALITIES")
display(
    companies["kommuneNavn"]
    .value_counts(dropna=False)
    .head(20)
    .to_frame("Count")
)

COMPANY STATUS


,Count
Status,
Ophørt,934018
Aktiv,468717
NORMAL,390963
OPLØSTEFTERKONKURS,130507
OPLØSTEFTERERKLÆRING,106125
TVANGSOPLØST,85157
OPLØSTEFTERFRIVILLIGLIKVIDATION,57451
OPLØSTEFTERFUSION,41027
SLETTET,12314



MOST COMMON INDUSTRIES


,Count
Branche,
Ikke-finansielle holdingselskaber,176456
Uoplyst,111084
Andre organisationers og foreningers aktiviteter i.a.n.,104874
NaN,67894
Andre organisationer og foreninger i.a.n.,62711
Udlejning af erhvervsejendomme,61077
Virksomhedsrådgivning og anden ledelsesrådgivning,32159
"Dyrkning af korn (undtagen ris), bælgfrugter og olieholdige frø",30261
Computerprogrammering,29153



MOST COMMON MUNICIPALITIES


,Count
kommuneNavn,
KØBENHAVN,335487
AARHUS,124627
AALBORG,72601
ODENSE,64688
NaN,58501
FREDERIKSBERG,49367
GENTOFTE,46417
VEJLE,39545
ESBJERG,34158


####**Første observationer**
* Datasættet indeholder 2.250.983 virksomheder, og hvert CVR-nummer optræder kun én gang.
* Der mangler brancheoplysninger for ca. 3 % af virksomhederne.
* Virksomhedsstatus er ikke helt ensartet angivet, fx forskelle i brugen af store og små bogstaver.
* Branchebetegnelserne ser ud til at indeholde forskellige eller overlappende navngivninger.
  * Dette bør undersøges nærmere, inden de eventuelt standardiseres.

###Finance

In [11]:
finance = pd.read_csv(finance_path, low_memory=False)

finance.columns = finance.columns.str.strip()

print("Shape:", finance.shape)

print("\nColumns:")
print(finance.columns.tolist())

print("\nDatatypes:")
display(finance.dtypes)

print("\nMemory usage:")
print(f"{finance.memory_usage(deep=True).sum() / 1024**2:.1f} MB")

Shape: (3763613, 8)

Columns:
['CVR', 'LocalName', 'Value', 'Context', 'SlutDato', 'InstantDate', 'DocType', 'DKLabel']

Datatypes:


,0
CVR,int64
LocalName,object
Value,float64
Context,object
SlutDato,object
InstantDate,object
DocType,object
DKLabel,object



Memory usage:
1334.0 MB


In [12]:
print("ANTAL UNIKKE VÆRDIER")
display(finance.nunique().to_frame("Antal unikke"))

print("\nMANGLENDE VÆRDIER")
missing_finance = pd.DataFrame({
    "Antal manglende": finance.isna().sum(),
    "Procent": (finance.isna().mean() * 100).round(2)
})
display(missing_finance)

print("\nANTAL UNIKKE VIRKSOMHEDER")
print(f"{finance['CVR'].nunique():,}".replace(",", "."))

print("\nDOKUMENTTYPER")
display(finance["DocType"].value_counts(dropna=False))

print("\n20 MEST FOREKOMMENDE REGNSKABSPOSTER")
display(
    finance["LocalName"]
    .value_counts(dropna=False)
    .head(20)
    .to_frame("Antal")
)

ANTAL UNIKKE VÆRDIER


,Antal unikke
CVR,356446
LocalName,17
Value,1672483
Context,399
SlutDato,165
InstantDate,165
DocType,1
DKLabel,17



MANGLENDE VÆRDIER


,Antal manglende,Procent
CVR,0,0.00
LocalName,0,0.00
Value,0,0.00
Context,0,0.00
SlutDato,0,0.00
InstantDate,2237220,59.44
DocType,0,0.00
DKLabel,0,0.00



ANTAL UNIKKE VIRKSOMHEDER
356.446

DOKUMENTTYPER


,count
DocType,
AARSRAPPORT,3763613



20 MEST FOREKOMMENDE REGNSKABSPOSTER


,Antal
LocalName,
ProfitLoss,365570
Equity,365437
Assets,365152
ShorttermLiabilitiesOtherThanProvisions,352052
ProfitLossFromOrdinaryActivitiesBeforeTax,347818
AverageNumberOfEmployees,340693
CashAndCashEquivalents,327100
GrossProfitLoss,310223
ProfitLossFromOrdinaryOperatingActivities,287545


In [13]:
financial_items = (
    finance.groupby(["LocalName", "DKLabel"])
    .agg(
        Antal=("Value", "size"),
        Virksomheder=("CVR", "nunique"),
        Mangler_InstantDate=("InstantDate", lambda x: x.isna().sum()),
        Mangler_SlutDato=("SlutDato", lambda x: x.isna().sum())
    )
    .reset_index()
    .sort_values("Antal", ascending=False)
)

financial_items["Dækning_pct"] = (
    financial_items["Virksomheder"] / finance["CVR"].nunique() * 100
).round(1)

display(financial_items)

,LocalName,DKLabel,Antal,Virksomheder,Mangler_InstantDate,Mangler_SlutDato,Dækning_pct
11,ProfitLoss,Resultat,365570,356396,365570,0,100.0
3,Equity,Egenkapital,365437,356410,0,0,100.0
0,Assets,Balancesum,365152,356412,0,0,100.0
16,ShorttermLiabilitiesOtherThanProvisions,Kortsigtede forpligtelser,352052,343605,0,0,96.4
12,ProfitLossFromOrdinaryActivitiesBeforeTax,Resultat før skat,347818,339533,347818,0,95.3
1,AverageNumberOfEmployees,Antal ansatte,340693,324972,340693,0,91.2
2,CashAndCashEquivalents,Likvide beholdninger,327100,319087,0,0,89.5
5,GrossProfitLoss,Brutto fortjeneste,310223,303202,310223,0,85.1
13,ProfitLossFromOrdinaryOperatingActivities,Resultat af primær drift,287545,280821,287545,0,78.8
10,OtherFinanceIncome,Andre finansielle indtægter,268933,262411,268933,0,73.6


In [14]:
finance["SlutDato_dt"] = pd.to_datetime(finance["SlutDato"], errors="coerce")
finance["InstantDate_dt"] = pd.to_datetime(finance["InstantDate"], errors="coerce")

print("SlutDato:")
print("Fra:", finance["SlutDato_dt"].min())
print("Til:", finance["SlutDato_dt"].max())

print("\nInstantDate:")
print("Fra:", finance["InstantDate_dt"].min())
print("Til:", finance["InstantDate_dt"].max())

SlutDato:
Fra: 2025-01-03 00:00:00
Til: 2026-07-31 00:00:00

InstantDate:
Fra: 2025-01-03 00:00:00
Til: 2026-07-31 00:00:00


In [15]:
profit_counts = (
    finance[finance["LocalName"] == "ProfitLoss"]
    .groupby("CVR")
    .size()
)

print("Antal ProfitLoss-observationer pr. virksomhed:")
display(profit_counts.value_counts().sort_index())

example_cvr = profit_counts[profit_counts > 1].index[0]

print(f"\nEksempel - CVR {example_cvr}:")
display(
    finance[
        (finance["CVR"] == example_cvr) &
        (finance["LocalName"] == "ProfitLoss")
    ].sort_values("SlutDato")
)

Antal ProfitLoss-observationer pr. virksomhed:


,count
1,347346
2,8936
3,104
4,10



Eksempel - CVR 10007062:


,CVR,LocalName,Value,Context,SlutDato,InstantDate,DocType,DKLabel,SlutDato_dt,InstantDate_dt
2951886,10007062,ProfitLoss,1923.0,c0,2025-04-30,NaN,AARSRAPPORT,Resultat,2025-04-30,NaT
2951875,10007062,ProfitLoss,114077.0,c0,2026-04-30,NaN,AARSRAPPORT,Resultat,2026-04-30,NaT


In [16]:
print("MEST FOREKOMMENDE CONTEXT-VÆRDIER")
display(
    finance["Context"]
    .value_counts()
    .head(20)
    .to_frame("Antal")
)

print("\nEKSEMPEL PÅ ET CVR MED FLERE REGNSKABSÅR")
display(
    finance[finance["CVR"] == 10007062]
    .sort_values(["SlutDato", "LocalName"])
)

MEST FOREKOMMENDE CONTEXT-VÆRDIER


,Antal
Context,
c1,664078
c0,433807
c11,254497
c4,229717
ctx-1,194476
ID_0,180133
c45,169561
c49,166236
c40,161104



EKSEMPEL PÅ ET CVR MED FLERE REGNSKABSÅR


,CVR,LocalName,Value,Context,SlutDato,InstantDate,DocType,DKLabel,SlutDato_dt,InstantDate_dt
2951887,10007062,Assets,1595025.0,c5,2025-04-30,2025-04-30,AARSRAPPORT,Balancesum,2025-04-30,2025-04-30
2951880,10007062,AverageNumberOfEmployees,0.0,c0,2025-04-30,NaN,AARSRAPPORT,Antal ansatte,2025-04-30,NaT
2951888,10007062,CashAndCashEquivalents,35529.0,c5,2025-04-30,2025-04-30,AARSRAPPORT,Likvide beholdninger,2025-04-30,2025-04-30
2951889,10007062,Equity,1556790.0,c5,2025-04-30,2025-04-30,AARSRAPPORT,Egenkapital,2025-04-30,2025-04-30
2951881,10007062,GrossProfitLoss,-31113.0,c0,2025-04-30,NaN,AARSRAPPORT,Brutto fortjeneste,2025-04-30,NaT
2951884,10007062,OtherFinanceExpenses,182.0,c0,2025-04-30,NaN,AARSRAPPORT,Andre finansielle udgifter,2025-04-30,NaT
2951883,10007062,OtherFinanceIncome,434.0,c0,2025-04-30,NaN,AARSRAPPORT,Andre finansielle indtægter,2025-04-30,NaT
2951886,10007062,ProfitLoss,1923.0,c0,2025-04-30,NaN,AARSRAPPORT,Resultat,2025-04-30,NaT
2951885,10007062,ProfitLossFromOrdinaryActivitiesBeforeTax,1923.0,c0,2025-04-30,NaN,AARSRAPPORT,Resultat før skat,2025-04-30,NaT
2951882,10007062,ProfitLossFromOrdinaryOperatingActivities,-31113.0,c0,2025-04-30,NaN,AARSRAPPORT,Resultat af primær drift,2025-04-30,NaT


####**Første observationer**
* Datasættet indeholder 3.763.613 observationer fra 356.446 virksomheder.
* Der findes 17 forskellige regnskabsposter, og dækningen varierer betydeligt mellem dem.
* Resultat, egenkapital og balancesum findes for næsten alle virksomheder med regnskabsdata.
* Omsætning findes kun for ca. 6,7 % af virksomhederne og er derfor mindre egnet som et generelt benchmark.
* Manglende værdier i InstantDate følger typen af regnskabspost og ser derfor ud til at være strukturelle frem for tilfældige.
* Regnskabsobservationerne har slutdatoer fra januar 2025 til juli 2026.
* De fleste virksomheder har ét regnskab i datasættet, mens en mindre gruppe har flere regnskabsperioder. Disse observationer er derfor ikke nødvendigvis dubletter.

###Sammenhæng mellem Companies og Finance

In [17]:
finance_cvrs = set(finance["CVR"].unique())
companies_cvrs = set(companies["Cvr"].unique())

matching_cvrs = finance_cvrs.intersection(companies_cvrs)
finance_only = finance_cvrs - companies_cvrs

print(f"Unikke CVR i Finance: {len(finance_cvrs):,}".replace(",", "."))
print(f"Unikke CVR i Companies: {len(companies_cvrs):,}".replace(",", "."))
print(f"CVR som findes i begge: {len(matching_cvrs):,}".replace(",", "."))
print(f"CVR fra Finance som ikke findes i Companies: {len(finance_only):,}".replace(",", "."))

match_pct = len(matching_cvrs) / len(finance_cvrs) * 100
print(f"\nMatch-rate for Finance: {match_pct:.2f}%")

Unikke CVR i Finance: 356.446
Unikke CVR i Companies: 2.250.983
CVR som findes i begge: 356.446
CVR fra Finance som ikke findes i Companies: 0

Match-rate for Finance: 100.00%


Finance kan kobles til Companies via CVR-nummer. Alle 356.446 virksomheder i Finance findes i Companies, svarende til en match-rate på 100 %. Companies kan derfor bruges til at berige regnskabsdata med bl.a. virksomhedsnavn, branche, kommune og status.

#Step 2: Data Preparation